### PR TITLE
Reduce async overhead

### PR DESCRIPTION
**Pull request by pentp**
Removed async overhead from many functions (state machines/closures/continuations/exception re-throwing).

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 1]
```diff
@@ -498,7 +498,7 @@ public virtual Container GetContainer(string databaseId, string containerId)
                 using (ResponseMessage readResponse = await database.ReadStreamAsync(
                     diagnosticsContext: diagnostics,
                     requestOptions: requestOptions,
-                    cancellationToken: cancellationToken))
+                    cancellationToken: cancellationToken).ConfigureAwait(false))
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:501]: This shouldn't be necessary all operations go through the [OperationHelperAsync ](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/d48fb274e84f00a965dacd53028f3623be457365/Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs#L409) which already sets the configure await to false. What is the point to adding ConfigureAwait to all the awaited calls?

File: Microsoft.Azure.Cosmos/src/Handler/AbstractRetryHandler.cs [hunk 2]
```diff
@@ -13,32 +13,54 @@ namespace Microsoft.Azure.Cosmos.Handlers
...
         {
-            IDocumentClientRetryPolicy retryPolicyInstance = await this.GetRetryPolicyAsync(request);
+            IDocumentClientRetryPolicy retryPolicyInstance = this.GetRetryPolicy(request);
             request.OnBeforeSendRequestActions += retryPolicyInstance.OnBeforeSendRequest;
 
             try
             {
-                return await RetryHandler.ExecuteHttpRequestAsync(
-                    callbackMethod: () =>
+                while (true)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Handler/AbstractRetryHandler.cs:27]: What's the reason for removing the ExecuteHttpRequestAsync helper?

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs (hunk 1 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs:501]: `OperationHelperAsync` has no effect when a custom `TaskScheduler` is used, so this code runs potentially in a custom scheduler context - this is totally fine, but the continuation (the code after `await`) should be scheduled to the thread pool.

File: Microsoft.Azure.Cosmos/src/Handler/AbstractRetryHandler.cs (hunk 2 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/Handler/AbstractRetryHandler.cs:27]: It's a single-use helper that doesn't help much and just causes overhead here (at least 3 delegates, state machine, closure and task allocations + execution overhead).

File: Microsoft.Azure.Cosmos/src/TimerWheel/TimerWheelTimerCore.cs [hunk 3]
```diff
@@ -56,7 +55,7 @@ public override bool CancelTimer()
 
         public override bool FireTimeout()
         {
-            return this.taskCompletionSource.TrySetResult(TimerWheelTimerCore.completedObject);
+            return this.taskCompletionSource.TrySetResult(null);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/TimerWheel/TimerWheelTimerCore.cs:58]: Please add tests if the configure await is needed for custom schedulers.

Please run the perf results showing that all these changes are actually improvements.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs [hunk 4]
```diff
@@ -136,35 +134,30 @@ public void Dispose()
...
-            this.timerTask = this.currentTimer.StartTimerAsync().ContinueWith((task) =>
+            _ = this.currentTimer.StartTimerAsync().ContinueWith((task) =>
             {
-                if (task.IsCompleted)
-                {
-                    this.DispatchTimer();
-                }
-            }, this.cancellationTokenSource.Token);
+                this.DispatchTimer();
+            }, this.cancellationTokenSource.Token, TaskContinuationOptions.NotOnCanceled, TaskScheduler.Default);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:151]: Why is this explicitly on the TaskScheduler.Default?

File: Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/FeedProcessing/FeedEstimatorCore.cs [hunk 5]
```diff
@@ -27,24 +27,19 @@ public override async Task RunAsync(CancellationToken cancellationToken)
...
-                    if (cancellationToken.IsCancellationRequested)
-                        throw;
-
                     Extensions.TraceException(new Exception("exception within estimator", canceledException));
 
                     // ignore as it is caused by client
                 }
 
-                await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
+                await Task.Delay(this.monitoringDelay, cancellationToken).ConfigureAwait(false);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/FeedProcessing/FeedEstimatorCore.cs:42]: Good catch

File: Microsoft.Azure.Cosmos/src/ClientExtensions.cs [hunk 6]
```diff
@@ -1,63 +0,0 @@
...
-        public static async Task<HttpResponseMessage> GetAsync(this HttpClient client,
-            Uri uri,
-            INameValueCollection additionalHeaders = null,
-            CancellationToken cancellationToken = default(CancellationToken))
-        {
-            if (uri == null) throw new ArgumentNullException("uri");
-
-            // GetAsync doesn't let clients to pass in additional headers. So, we are
-            // internally using SendAsync and add the additional headers to requestMessage. 
-            using (HttpRequestMessage requestMessage = new HttpRequestMessage(HttpMethod.Get, uri))
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ClientExtensions.cs:25]: @j82w Does this file get used with your HttpClient refactoring?

File: Microsoft.Azure.Cosmos/src/ConnectionPolicy.cs [hunk 7]
```diff
@@ -85,16 +79,6 @@ public void SetPreferredLocations(IReadOnlyList<string> regions)
             }
         }
 
-        /// <summary>
-        /// Gets or sets the maximum number of concurrent fanout requests sent to the Azure Cosmos DB service.
-        /// </summary>
-        /// <value>Default value is 32.</value>
-        internal int MaxConcurrentFanoutRequests
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ConnectionPolicy.cs:92]: +@j82w This is altering internal contract, these APIs might be used internally. The fact that they have no reference might not mean they are not used.

File: Microsoft.Azure.Cosmos/src/DocumentClientWithUriParameters.cs [hunk 8]
```diff
@@ -86,24 +86,16 @@ public Task<ResourceResponse<DocumentCollection>> CreateDocumentCollectionIfNotE
             {
                 return await this.ReadDocumentCollectionAsync(documentCollectionUri, options);
             }
-            catch (DocumentClientException dce)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/DocumentClientWithUriParameters.cs:89]: These are old files, changes here do not impact anything

File: Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs [hunk 9]
```diff
@@ -89,22 +89,11 @@ internal static async Task<DocumentServiceResponse> ParseResponseAsync(HttpRespo
             using (responseMessage)
             {
                 IClientSideRequestStatistics requestStatistics = request?.RequestContext?.ClientRequestStatistics;
-                if ((int)responseMessage.StatusCode < 400)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs:92]: Can we scope this PR to adding ConfigureAwait? These code refactorings are impacting files which are synced with internal sources. The outcome (aside from the ConfigureAwait changes) seems to be just refactoring (in fact in this file it is removing the `serializerSettings` parameter).

**Review by ealsur — COMMENTED** [review]
Can we scope the PR just to the ConfigureAwait adding or the async removal and not do all the other code refactorings, file reorganizations, and renamings?

**Comment by YohanSciubukgian** [comment]
Did you consider using ValueTask to reduce async overhead also ? (Not sure if it match your SDK rule contraints)
https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/

It as been widely used by the dotnet team and have added significant performance improvements by removing the async overhead.

File: Microsoft.Azure.Cosmos/src/DocumentClientWithUriParameters.cs (hunk 8 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/DocumentClientWithUriParameters.cs:89]: Right, it does look like there are no callers for these functions. But why are these files then still included in the project?

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs (hunk 4 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:151]: Because by default `TaskScheduler.Current` is used (which could be custom scheduler).

**Comment by pentp** [comment]
I scoped this PR just to async overhead removal.

**Comment by j82w** [comment]
> Did you consider using ValueTask to reduce async overhead also ? (Not sure if it match your SDK rule contraints)
> https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/
> 
> It as been widely used by the dotnet team and have added significant performance improvements by removing the async overhead.

@YohanSciubukgian we did some initial testing for converting all the caches to ValueTask and didn't see that big of an improvement in performance. We decided to prioritize other changes based on profiling that would have a larger impact.

File: Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs [hunk 10]
```diff
@@ -254,13 +254,13 @@ public DataEncryptionKeyContainerCore(CosmosDataEncryptionKeyProvider dekProvide
...
 
-        private async Task<ItemResponse<DataEncryptionKeyProperties>> ReadInternalAsync(
+        private Task<ItemResponse<DataEncryptionKeyProperties>> ReadInternalAsync(
             string id,
             ItemRequestOptions requestOptions,
             CosmosDiagnosticsContext diagnosticsContext,
             CancellationToken cancellationToken)
         {
-            return await this.DekProvider.Container.ReadItemAsync<DataEncryptionKeyProperties>(
+            return this.DekProvider.Container.ReadItemAsync<DataEncryptionKeyProperties>(
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs:263]: Was there a tool you used to catch all these? If so can you let us know and maybe we can have it run as part of the gates? Thanks :)

Comment (pentp) [reply Microsoft.Azure.Cosmos.Encryption/src/DataEncryptionKeyContainerCore.cs:263]: No tool, just find "return await"

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs [hunk 11]
```diff
@@ -1368,11 +1368,7 @@ public void Dispose()
                 this.storeClientFactory = null;
             }
 
-            if (this.AddressResolver != null)
-            {
-                this.AddressResolver.Dispose();
-                this.AddressResolver = null;
-            }
+            this.AddressResolver = null;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/DocumentClient.cs:1371]: This should still call Dispose, right?

File: Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs (hunk 9 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs:92]: Can we undo this file? This change is not related to async/Tasks

File: Microsoft.Azure.Cosmos/src/Handler/RequestHandler.cs [hunk 12]
```diff
@@ -38,9 +38,17 @@ public abstract class RequestHandler
                 throw new ArgumentNullException(nameof(this.InnerHandler));
             }
 
-            using (request.DiagnosticsContext.CreateRequestHandlerScopeScope(this.InnerHandler))
+            IDisposable scope = request.DiagnosticsContext.CreateRequestHandlerScopeScope(this.InnerHandler);
+            try
             {
-                return await this.InnerHandler.SendAsync(request, cancellationToken);
+                Task<ResponseMessage> task = this.InnerHandler.SendAsync(request, cancellationToken);
+                _ = task.ContinueWith((_, state) => ((IDisposable)state).Dispose(), scope, default, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Handler/RequestHandler.cs:45]: Does removing the `await` make such a big impact to have this more complex code around?

File: Microsoft.Azure.Cosmos/src/Routing/AsyncCache.cs [hunk 13]
```diff
@@ -209,46 +204,24 @@ public void Clear()
...
             oldValues.Clear();
         }
 
-        /// <summary>
-        /// Runs a background task that will started refreshing the cached value for a given key.
-        /// This observes the same logic as GetAsync - a running value will still take precedence over a call to this.
-        /// </summary>
-        /// <param name="key">Key.</param>
-        /// <param name="singleValueInitFunc">Generator function.</param>
-        public void BackgroundRefreshNonBlocking(TKey key, Func<Task<TValue>> singleValueInitFunc)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/AsyncCache.cs:225]: This is removing this API because it has no references?

File: Microsoft.Azure.Cosmos/src/Routing/GatewayAddressCache.cs [hunk 14]
```diff
@@ -515,11 +515,6 @@ await ClientExtensions.ParseResponseAsync(httpResponseMessage))
             }
         }
 
-        public void Dispose()
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/GatewayAddressCache.cs:518]: How is these changes related to async/Task? What is the reason for their removal?

File: Microsoft.Azure.Cosmos/src/Routing/GlobalAddressResolver.cs [hunk 15]
```diff
@@ -153,14 +153,6 @@ private IAddressResolver GetAddressResolver(DocumentServiceRequest request)
             return this.GetOrAddEndpoint(endpoint).AddressResolver;
         }
 
-        public void Dispose()
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Routing/GlobalAddressResolver.cs:156]: Why are these being removed?

File: Microsoft.Azure.Cosmos/src/SecureStringHMACSHA256Helper.cs [hunk 16]
```diff
@@ -53,20 +53,19 @@ public void Dispose()
             GC.SuppressFinalize(this);
         }
 
+        ~SecureStringHMACSHA256Helper() => this.Dispose(false);
+
         private void Dispose(bool disposing)
         {
-            if (disposing)
+            if (this.algorithmHandle != IntPtr.Zero)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/SecureStringHMACSHA256Helper.cs:60]: No need to check for `disposing`?

File: Microsoft.Azure.Cosmos/src/Handler/RequestHandler.cs (hunk 12 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Handler/RequestHandler.cs:45]: Please do not use ContinueWith unless you have proof that it is absolutely needed. Doing the continue with causes all the exceptions to be aggregate exceptions and increases code complexity.

File: Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs (hunk 9 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs:92]: This is just removing code duplication, but `serializerSettings` are there as before.

File: Microsoft.Azure.Cosmos/src/DocumentClient.cs (hunk 11 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/DocumentClient.cs:1371]: No, but I'll move the finalizer related fixes to a separate PR

File: Microsoft.Azure.Cosmos/src/Handler/RequestHandler.cs (hunk 12 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/Handler/RequestHandler.cs:45]: This function is called repeatedly on each request for each handler so it's worth effort here, especially when exceptions are involved because the current async/await causes this helper function to be in the continuation chain in between each handler.

In this case the `ContinueWith` actually helps with exception handling - now this helper `SendAsync` function returns the inner task directly without wrapping it and the `ContinueWith` just does scope disposal and is transparent to this functions's callers.

File: Microsoft.Azure.Cosmos/src/Routing/AsyncCache.cs (hunk 13 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/Routing/AsyncCache.cs:225]: Yes

File: Microsoft.Azure.Cosmos/src/SecureStringHMACSHA256Helper.cs (hunk 16 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/SecureStringHMACSHA256Helper.cs]: Bot the disposing (always true, no finalizer defined) and `algorithmHandle != null` (always true) check were incorrect, but I'll create a separate PR for this.

File: Microsoft.Azure.Cosmos/src/Handler/RequestHandler.cs (hunk 12 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Handler/RequestHandler.cs:45]: The only problem with this is it breaks the diagnostic logic. The scope is going to measure how long it took to create the task rather than the entire operation time.

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Handler/RequestHandler.cs:45]: The only problem is the diagnostics.  The scope was measuring the time for the InnerHandler.SendAsync took. The current logic only measures the time it to return the task..

Comment (pentp) [reply Microsoft.Azure.Cosmos/src/Handler/RequestHandler.cs:45]: No, the new logic also measures the same duration - it's competed only when the task completes (or if the InnerHandler throws synchronously) like before.

File: Microsoft.Azure.Cosmos/src/Handler/TransportHandler.cs [hunk 17]
```diff
@@ -88,14 +88,12 @@ public TransportHandler(CosmosClient client)
             IStoreModel storeProxy = this.client.DocumentClient.GetStoreProxy(serviceRequest);
             using (request.DiagnosticsContext.CreateScope(storeProxy.GetType().FullName))
             {
+                DocumentServiceResponse response = await storeProxy.ProcessMessageAsync(serviceRequest, cancellationToken);
                 if (request.OperationType == OperationType.Upsert)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Handler/TransportHandler.cs:92]: Won't explicit if-else enable removing async-await from this function?

File: Microsoft.Azure.Cosmos/src/Handler/RetryHandler.cs [hunk 18]
```diff
@@ -25,13 +24,13 @@ public RetryHandler(CosmosClient client)
             this.client = client;
         }
 
-        internal override Task<IDocumentClientRetryPolicy> GetRetryPolicyAsync(RequestMessage request)
+        internal override IDocumentClientRetryPolicy GetRetryPolicy(RequestMessage request)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Handler/RetryHandler.cs:27]: hahah thanks. 

Pre-mature designs are evil.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs [hunk 19]
```diff
@@ -136,35 +134,30 @@ public void Dispose()
...
                 this.congestionControlTimer = null;
-                this.congestionControlTask = null;
             }
         }
 
         private void ResetTimer()
         {
             this.currentTimer = this.timerWheel.CreateTimer(BatchAsyncStreamer.batchTimeout);
-            this.timerTask = this.currentTimer.StartTimerAsync().ContinueWith((task) =>
+            _ = this.currentTimer.StartTimerAsync().ContinueWith((task) =>
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:148]: Do Tasks GCed before completion? 

We were unsure of this behavior and guess it why this defensive pattern.
/cc: @ealsur , @FabianMeiswinkel , @j82w

**Comment by j82w** [comment]
@pentp is there anything I can do to help get this conflicts resolved so this can get merged?

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs [hunk 20]
```diff
@@ -136,35 +134,30 @@ public void Dispose()
...
             {
-                if (task.IsCompleted)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:154]: Wow, thanks.

**Review by ealsur — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs (hunk 19 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:148]: Yeah, basically that was the reason we were maintaining the reference.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs (hunk 4 above)
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:151]: What if the Task failed, it would count as NotOnCanceled, right?

Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:151]: Specifying TaskSchedulers have some side-effects. 

@ealsur , can you please take a quick look?
ex: Actor based single scheduler, Compute CPU tracking etc...

@pentp how does the scheduler matter?

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs (hunk 19 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:148]: I guess since we pass the cancellation, and we cancel on disposing, it would cancel the Tasks

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs (hunk 20 above)
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:154]: @ealsur why is this check needed. Won't ContiueWith be for all terminal conditions?

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:154]: Because the ContinueWith, unless specified, will also fire if the task failed or was cancelled. So with this check, we don't fire the Dispatch if the task was canceled due to the Stream disposing and cancelling the continuation token source

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs (hunk 4 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:151]: From the Orleans docs: https://dotnet.github.io/orleans/Documentation/grains/external_tasks_and_grains.htm, I think what @pentp might be trying to do is avoid any task continuation going to `TaskScheduler.Current` and specifically send it to the Default. I am not familiar with Compute to know if this might affect them, @visridha do you know?

**Review by ealsur — COMMENTED** [review]
Picked accidental Approve, meant to use Comment.

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs (hunk 1 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs]: I think this is related to https://dotnet.github.io/orleans/Documentation/grains/external_tasks_and_grains.htm

File: Microsoft.Azure.Cosmos/src/Handler/TransportHandler.cs (hunk 17 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/Handler/TransportHandler.cs:92]: Not without changing the way the diagnostics scope is handled. It could be done using `ContinueWith`, but it's probably not worth the complexity here.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs (hunk 19 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:148]: The task won't get GCed while it's running or queued on some task scheduler.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs (hunk 4 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:151]: Yes.

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs (hunk 20 above)
Comment (pentp) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:154]: This `if (task.IsCompleted)` was actually a no-op before (always true, even if the task failed or was canceled). I think the intention here was not to run the continuation if the task was canceled, so that's why I added `TaskContinuationOptions.NotOnCanceled`

File: Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs (hunk 4 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs:151]: @pentp can you separate the task scheduler changes into a new PR? Let this PR just be for the async awaits.

Comment (pentp) [reply Microsoft.Azure.Cosmos/src/Batch/BatchAsyncStreamer.cs]: Rolled back the TaskScheduler changes

**Comment by pentp** [comment]
I changed a couple of callsites back to `TaskScheduler.Current` based on https://github.com/Azure/azure-cosmos-dotnet-v3/pull/1839#discussion_r497506839 so that this PR does not change continuation scheduling behavior.

**Comment by j82w** [comment]
@pentp can you fix the merge conflicts so we can get this merged?

**Comment by pentp** [comment]
Fixed the merge conflicts.

**Comment by pentp** [comment]
Could we get this merged?

**Comment by j82w** [comment]
@pentp can you investigate the test failures?
```
X RunObserver_ShouldCancelTasks_WhenTokenCanceled [134ms]
  Error Message:
   Test method Microsoft.Azure.Cosmos.ChangeFeed.Tests.PartitionSupervisorTests.RunObserver_ShouldCancelTasks_WhenTokenCanceled threw exception: 
System.Threading.Tasks.TaskCanceledException: A task was canceled.
  Stack Trace:
      at Microsoft.Azure.Cosmos.ChangeFeed.FeedManagement.PartitionSupervisorCore`1.RunAsync(CancellationToken shutdownToken) in D:\a\1\s\Microsoft.Azure.Cosmos\src\ChangeFeedProcessor\FeedManagement\PartitionSupervisorCore.cs:line 78
   at Microsoft.Azure.Cosmos.ChangeFeed.Tests.PartitionSupervisorTests.RunObserver_ShouldCancelTasks_WhenTokenCanceled() in D:\a\1\s\Microsoft.Azure.Cosmos\tests\Microsoft.Azure.Cosmos.Tests\ChangeFeed\PartitionSupervisorTests.cs:line 63
```
